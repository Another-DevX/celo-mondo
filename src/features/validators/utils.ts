import { ValidatorGroup, ValidatorStatus } from 'src/features/validators/types';
import { eqAddressSafe } from 'src/utils/addresses';
import { fromFixidity } from 'src/utils/numbers';
import { toTitleCase, trimToLength } from 'src/utils/strings';

export function findGroup(groups?: ValidatorGroup[], address?: Address) {
  if (!groups || !address) return undefined;
  return groups.find((g) => eqAddressSafe(g.address, address));
}

export function cleanGroupName(name: string) {
  return trimToLength(toTitleCase(name.replace(/group|Group/g, '').replace(/[-_]/g, ' ')), 20);
}

export function isElected(group: ValidatorGroup) {
  return Object.values(group.members).some((m) => m.status === ValidatorStatus.Elected);
}

export function getGroupStats(group?: ValidatorGroup) {
  if (!group) return { numMembers: 0, numElected: 0, avgScore: 0 };
  const members = Object.values(group.members);
  const electedMembers = members.filter((m) => m.status === ValidatorStatus.Elected);
  // TODO: group.score is between 0 and 1
  const avgScore = fromFixidity(group.score) * 100;
  return { numMembers: members.length, numElected: electedMembers.length, avgScore };
}
