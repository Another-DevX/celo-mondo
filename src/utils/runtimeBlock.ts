export default function getRuntimeBlock(): { blockNumber: bigint } | { blockTag: 'latest' } {
  if (process.env.NODE_ENV !== 'production' && process.env.NEXT_PUBLIC_FORK_BLOCK_NUMBER) {
    return { blockNumber: BigInt(process.env.NEXT_PUBLIC_FORK_BLOCK_NUMBER) };
  }
  return { blockTag: 'latest' };
}
