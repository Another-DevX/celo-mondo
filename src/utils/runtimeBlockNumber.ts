export default function getRuntimeBlockNumber() {
  if (process.env.NODE_ENV !== 'production') {
    return process.env.NEXT_PUBLIC_FORK_BLOCK_NUMBER
      ? BigInt(process.env.NEXT_PUBLIC_FORK_BLOCK_NUMBER)
      : undefined;
  }
  return undefined;
}
