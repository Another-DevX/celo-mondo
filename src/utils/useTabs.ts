import { useRouter, useSearchParams } from 'next/navigation';
import { useCallback, useState } from 'react';

export default function useTabs<
  Tabs extends readonly string[],
  Tab extends Tabs[number] = Tabs[number],
>(defaultTab: Tab) {
  const router = useRouter();
  const params = useSearchParams();
  const initialTab = (params?.get('tab') as Tab) || defaultTab;
  const [tab, setTab] = useState<Tabs[number]>(initialTab);

  const onTabChange = useCallback(
    (tab: Tab) => {
      setTab(tab);
      if (params?.get('tab') !== tab) {
        router.push('?tab=' + tab);
      }
    },
    [router, params],
  );

  return {
    onTabChange,
    tab,
  };
}
