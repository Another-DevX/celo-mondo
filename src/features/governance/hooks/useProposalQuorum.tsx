import { governanceABI } from '@celo/abis';
import { useQuery } from '@tanstack/react-query';
import BigNumber from 'bignumber.js';
import { useMemo } from 'react';
import { useToastError } from 'src/components/notifications/useToastError';
import { Addresses } from 'src/config/contracts';
import { MergedProposalData } from 'src/features/governance/hooks/useGovernanceProposals';
import { Proposal } from 'src/features/governance/types';
import { FORK_BLOCK_NUMBER } from 'src/test/anvil/constants';
import { fromFixidity } from 'src/utils/numbers';
import { PublicClient } from 'viem';
import { usePublicClient, useReadContract } from 'wagmi';

interface ParticipationParameters {
  baseline: number;
  baselineFloor: number;
  baselineUpdateFactor: number;
  baselineQuorumFactor: number;
}

export function useProposalQuorum(propData?: MergedProposalData) {
  const participationParameters = useParticipationParameters();
  const thresholds = useThresholds(propData?.proposal);

  if (!propData || !propData.proposal) return;
  if (!thresholds || !participationParameters) return;

  // https://github.com/celo-org/celo-monorepo/blob/master/packages/protocol/contracts/governance/Governance.sol#L1567
  const quorumPct = new BigNumber(participationParameters.baseline).times(
    participationParameters.baselineQuorumFactor,
  );
  // https://github.com/celo-org/celo-monorepo/blob/master/packages/protocol/contracts/governance/Proposals.sol#L195-L211
  const quorumVotes = BigInt(
    quorumPct.times(propData.proposal.networkWeight.toString()).toFixed(0),
  );
  const maxThreshold = Math.max(...thresholds);
  return new BigNumber(quorumVotes.toString()).times(maxThreshold).toFixed(0);
}

function useParticipationParameters(): ParticipationParameters {
  const { data } = useReadContract({
    blockNumber: process.env.NODE_ENV === 'development' ? FORK_BLOCK_NUMBER : undefined,
    address: Addresses.Governance,
    abi: governanceABI,
    functionName: 'getParticipationParameters',
    query: {
      gcTime: Infinity,
      staleTime: 60 * 60 * 1000, // 1 hour
    },
  });

  return useMemo(
    () => ({
      baseline: fromFixidity(data?.[0]),
      baselineFloor: fromFixidity(data?.[1]),
      baselineUpdateFactor: fromFixidity(data?.[2]),
      baselineQuorumFactor: fromFixidity(data?.[3]),
    }),
    [data],
  );
}

function useThresholds(proposal?: Proposal) {
  const publicClient = usePublicClient();
  const { isLoading, error, data } = useQuery({
    queryKey: ['useThresholds', publicClient],
    queryFn: async () => {
      if (!publicClient || !proposal) return null;
      return await fetchThresholds(publicClient, proposal);
    },
  });
  useToastError(error, 'Error fetching proposal quorum data');

  return isLoading ? undefined : data;
}

async function fetchThresholds(publicClient: PublicClient, proposal: Proposal) {
  const txIds = new Array(Number(proposal.numTransactions)).fill(0).map((_, id) => id);

  // Extracting the base contract call avoids the following error:
  // Type instantiation is excessively deep and possibly infinite. ts(2589)
  const getProposalTransactionContract = {
    address: Addresses.Governance,
    abi: governanceABI,
    functionName: 'getProposalTransaction',
  } as const;

  const results = await publicClient?.multicall({
    allowFailure: false,
    blockNumber: process.env.NODE_ENV === 'development' ? FORK_BLOCK_NUMBER : undefined,
    contracts: txIds.map((txId: number) => ({
      ...getProposalTransactionContract,
      args: [proposal.id, txId],
    })),
  });

  if (!results) {
    return;
  }

  // Extracting the base contract call avoids the following error:
  // Type instantiation is excessively deep and possibly infinite. ts(2589)
  const getConstitutionContract = {
    address: Addresses.Governance,
    abi: governanceABI,
    functionName: 'getConstitution',
  } as const;

  const thresholds = await publicClient?.multicall({
    allowFailure: false,
    blockNumber: process.env.NODE_ENV === 'development' ? FORK_BLOCK_NUMBER : undefined,
    contracts: results.map(([_value, destination, data]) => {
      const functionId = extractFunctionSignature(data);
      return {
        ...getConstitutionContract,
        args: [destination, functionId],
      };
    }),
  });

  if (!thresholds) {
    return;
  }

  // https://github.com/celo-org/celo-monorepo/blob/master/packages/protocol/contracts/governance/Governance.sol#L1580-L1583
  return thresholds.map(fromFixidity);
}

/**
 *
 * @notice Extracts the first four bytes of a byte array.
 * https://github.com/celo-org/celo-monorepo/blob/master/packages/protocol/contracts/common/ExtractFunctionSignature.sol#L9
 */
function extractFunctionSignature(input: `0x${string}`): `0x${string}` {
  const data = Buffer.from(input.replace('0x', ''), 'hex');
  return `0x${data.subarray(0, 4).toString('hex')}`;
}
