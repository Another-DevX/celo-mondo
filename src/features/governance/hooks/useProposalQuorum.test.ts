import * as rq from '@tanstack/react-query';
import BigNumber from 'bignumber.js';
import { afterEach, beforeEach } from 'node:test';
import { publicArchiveClient } from 'src/test/anvil/utils';
import { describe, expect, it, vi } from 'vitest';
import * as module from './useProposalQuorum';

beforeEach(() => {
  vi.mock('wagmi', async (importActual) => ({
    ...(await importActual()),
    usePublicClient: () => publicArchiveClient,
    useReadContract: vi.fn(() => ({
      data: [
        60000000000000000000000n, // baseline
        50000000000000000000000n, // baselineFloor
        200000000000000000000000n, // baselineUpdateFactor
        500000000000000000000000n, // baselineQuorumFactor
      ],
    })),
  }));
  vi.mock('@tanstack/react-query', async (importActual) => ({
    ...(await importActual()),
    useQuery: vi.fn(({ data, error, isLoading }) => ({
      isLoading: false,
      isError: !error,
      error,
      data,
    })),
  }));
  vi.mock('src/components/notifications/useToastError', async (importActual) => ({
    ...(await importActual()),
    useToastError: () => {},
  }));
  vi.mock('src/features/governance/hooks/useProposalQuorum', async (importActual) => ({
    ...(await importActual()),
    useThresholds: vi.fn((data) => data),
    useParticipationParameters: vi.fn((data) => data),
  }));
  vi.mock('react', async (importActual) => ({
    ...(await importActual()),
    useMemo: vi.fn((x) => x()),
  }));
});

afterEach(() => {
  vi.clearAllMocks();
});

describe('extractFunctionSignature', () => {
  it('gets the four first bytes of a byte array', () => {
    expect(module.extractFunctionSignature('0x112233445566778899')).toBe('0x11223344');
    expect(module.extractFunctionSignature('1234567890' as `0x${string}`)).toBe('0x12345678');
  });
});

describe('fetchThresholds', () => {
  it('fetches the constitution threshholds properly #2', async () => {
    // Proposal for Creation of Celo Governance Guild's execution block - 1
    process.env.NEXT_PUBLIC_FORK_BLOCK_NUMBER = '0x1bb55ea';
    const expectedNumberOfTxs = 1n;
    const proposalId = 195;
    const spy = vi.spyOn(publicArchiveClient, 'multicall');
    await expect(
      module.fetchThresholds(publicArchiveClient, proposalId, expectedNumberOfTxs),
    ).resolves.toMatchSnapshot();
    expect(spy.mock.results[0].value).resolves.toHaveLength(Number(expectedNumberOfTxs));
    expect(spy.mock.calls.length).toBe(2);
  });

  it('fetches the constitution threshholds properly #2', async () => {
    // Enabling MENTO Governance's proposal block
    process.env.NEXT_PUBLIC_FORK_BLOCK_NUMBER = '0x1baa98c';
    const expectedNumberOfTxs = 50n;
    const proposalId = 196;
    const spy = vi.spyOn(publicArchiveClient, 'multicall');
    await expect(
      module.fetchThresholds(publicArchiveClient, proposalId, expectedNumberOfTxs),
    ).resolves.toMatchSnapshot();
    expect(spy.mock.results[0].value).resolves.toHaveLength(Number(expectedNumberOfTxs));
    expect(spy.mock.calls.length).toBe(2);
  });
});

describe('useParticipationParameters', () => {
  it('fetches the params from the governance contract', async () => {
    expect(module.useParticipationParameters()).toMatchInlineSnapshot(`
      {
        "baseline": 0.06,
        "baselineFloor": 0.05,
        "baselineQuorumFactor": 0.5,
        "baselineUpdateFactor": 0.2,
      }
    `);
  });
});

describe('useProposalQuorum', () => {
  it('maths properly', () => {
    const baseline = 0.06;
    const baselineQuorumFactor = 0.5;
    vi.spyOn(module, 'useParticipationParameters').mockReturnValue({
      baseline,
      baselineFloor: 0.05,
      baselineUpdateFactor: 0.2,
      baselineQuorumFactor,
    });
    // @ts-expect-error
    vi.spyOn(rq, 'useQuery').mockReturnValue({ data: [0.75, 0.1, 0.5] });
    const networkWeight = 200000000000000000000000000n;

    const quorumPct = baseline * baselineQuorumFactor;
    const quorumVotes = new BigNumber(networkWeight.toString()).times(quorumPct);
    const maxThreshold = quorumVotes.times(0.75 /* highest mocked threshold */);
    expect(
      // @ts-expect-error
      module.useProposalQuorum({ proposal: { id: 1, numTransactions: 3, networkWeight } }),
    ).toBe(BigInt(maxThreshold.toFixed(0)));
    expect(BigInt(maxThreshold.toFixed(0))).toMatchInlineSnapshot(`4500000000000000000000000n`);
  });
});
