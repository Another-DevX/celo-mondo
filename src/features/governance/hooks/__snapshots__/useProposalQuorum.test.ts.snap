// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`fetchThresholds > fetches the constitution threshholds properly #2 1`] = `
[
  0.5,
  0.5,
  0.5,
  0.9,
  0.9,
  0.9,
  0.9,
  0.5,
  0.8,
  0.9,
  0.5,
  0.9,
  0.9,
  0.9,
  0.8,
  0.9,
  0.5,
  0.5,
  0.5,
  0.5,
  0.5,
  0.5,
  0.5,
  0.5,
  0.5,
  0.5,
  0.5,
  0.5,
  0.5,
  0.5,
  0.5,
  0.5,
  0.5,
  0.5,
  0.5,
  0.9,
  0.9,
  0.5,
  0.5,
  0.5,
  0.5,
  0.8,
  0.9,
  0.5,
  0.5,
  0.5,
  0.5,
  0.5,
  0.5,
  0.5,
]
`;

exports[`fetchThresholds > fetches the constitution threshholds properly #2 2`] = `
[
  0.5,
  0.5,
  0.5,
  0.9,
  0.9,
  0.9,
  0.9,
  0.5,
  0.8,
  0.9,
  0.5,
  0.9,
  0.9,
  0.9,
  0.8,
  0.9,
  0.5,
  0.5,
  0.5,
  0.5,
  0.5,
  0.5,
  0.5,
  0.5,
  0.5,
  0.5,
  0.5,
  0.5,
  0.5,
  0.5,
  0.5,
  0.5,
  0.5,
  0.5,
  0.5,
  0.9,
  0.9,
  0.5,
  0.5,
  0.5,
  0.5,
  0.8,
  0.9,
  0.5,
  0.5,
  0.5,
  0.5,
  0.5,
  0.5,
  0.5,
]
`;
